horovod/spark/lightning/remote.py

```diff
@@ -121,6 +121,11 @@ def train(serialized_model):
             _val_steps_per_epoch = val_steps_per_epoch if val_steps_per_epoch else 1.0
 
             cuda_available = torch.cuda.is_available()
+            # We need to check all ranks have same device type for traning.
+            # Horovod doesn't support heterogeneous allreduce for gradients.
+            cuda_avail_list = hvd.allgather_object(cuda_available, name='device type')
+            assert cuda_avail_list.count(cuda_available) == hvd.size(), "All ranks don't have same device type!"
```

Comment:
One thing I forgot to point out: this should be an exception (probably `RuntimeError`) rather than an exception. Two reasons for this:

1. Exceptions can be caught, whereas assertions effectively terminate the program, so there's nothing the caller can do to recover.
2. Assertions can be removed by the interpreter at runtime when optimizers are enabled.

Assertions are primarily for debugging application code, whereas Exceptions are for errors at runtime.

Reply:
okay, I was thinking about raise assertion error instead of RuntimeError before.

Reply:
fixed.